File: plugins/dotori-vscode/src/hoverProvider.ts
import * as vscode from 'vscode';

/** Documentation for every keyword/value recognised by the dotori DSL. */
const KEYWORD_DOCS: Record<string, string> = {
    // ── Top-level blocks ────────────────────────────────────────────────────
    'project':      'Declares a build project. Usage: `project <Name> { ... }`',
    'package':      'Declares a package for distribution. Usage: `package { name = "..." version = "..." ... }`',

    // ── Project properties ──────────────────────────────────────────────────
    'type':         'Project output type.\n\nValues: `executable` | `static-library` | `shared-library` | `header-only`',
    'std':          'C++ standard to use.\n\nValues: `c++17` | `c++20` | `c++23`',
    'description':  'Human-readable description of the project.',
    'optimize':     'Optimization level.\n\nValues: `none` | `speed` | `size` | `full`',
    'debug-info':   'Debug information level.\n\nValues: `none` | `minimal` | `full`',
    'runtime-link': 'Link mode for the C/C++ runtime.\n\nValues: `static` | `dynamic`\n\n*Note*: iOS/tvOS/watchOS force `static`; UWP forces `dynamic`.',
    'libc':         'C runtime to use (Linux only).\n\nValues: `glibc` | `musl`',
    'stdlib':       'C++ standard library.\n\nValues: `libc++` | `libstdc++`',
    'lto':          'Enable Link Time Optimization.\n\nValues: `true` | `false`',
    'warnings':     'Warning level.\n\nValues: `none` | `default` | `all` | `extra`',
    'warnings-as-errors': 'Treat all warnings as errors.\n\nValues: `true` | `false`',
    'android-api-level':  'Minimum Android API level (e.g. `26`).',
    'macos-min':    'Minimum macOS version (e.g. `"12.0"`).',
    'ios-min':      'Minimum iOS version (e.g. `"15.0"`).',
    'tvos-min':     'Minimum tvOS version.',
    'watchos-min':  'Minimum watchOS version.',
    'manifest':     'Path to a Windows `.manifest` file to embed after linking.',

    // ── Blocks ──────────────────────────────────────────────────────────────
    'sources':       'Source file glob patterns.\n\nUsage:\n```\nsources {\n    include "src/**/*.cpp"\n    exclude "src/generated/**"\n}\n```',
    'modules':       'C++20 module source files (`.cppm`, `.ixx`).\n\nGenerates BMI files at build time.',
    'headers':       'Header search paths, classified as `public` or `private`.\n\nPublic headers are exposed to dependents.',
    'defines':       'Preprocessor definitions.\n\nUsage: `defines { "NDEBUG" "MY_MACRO=1" }`',
    'links':         'Libraries to link against.\n\nUsage: `links { "pthread" "dl" }`',
    'frameworks':    'Apple frameworks to link (macOS/iOS/tvOS/watchOS).\n\nUsage: `frameworks { "Foundation" "Metal" }`',
    'framework-paths': 'Paths to `.framework` or `.xcframework` bundles. Resolved at build time.',
    'resources':     'Windows resource files (`.rc`) to compile with `rc.exe`.',
    'compile-flags': 'Raw compiler flags appended after dotori-generated flags.',
    'link-flags':    'Raw linker flags appended after dotori-generated flags.',
    'emscripten-flags': 'Extra flags passed to `emcc` when targeting WebAssembly via Emscripten.',
    'dependencies':  'Project or package dependencies.\n\nExamples:\n```\ndependencies {\n    my-lib = { path = "../lib" }\n    fmt    = "10.2.0"\n    spdlog = { git = "https://github.com/gabime/spdlog", tag = "v1.13.0" }\n}\n```',
    'pch':           'Precompiled header configuration.\n\nUsage:\n```\npch {\n    header = "src/pch.h"\n    source = "src/pch.cpp"\n}\n```',
    'unity-build':   'Unity (jumbo) build settings.\n\nBatches multiple source files into a single compilation unit.\n\nUsage:\n```\nunity-build {\n    enabled    = true\n    batch-size = 8\n    exclude    { "src/main.cpp" }\n}\n```',
    'output':        'Directories for build artifacts (relative to project root).\n\nUsage:\n```\noutput {\n    binaries  = "bin/"    (* exe, dll/so/dylib *)\n    libraries = "lib/"    (* .a, import .lib  *)\n    symbols   = "pdb/"    (* .pdb, .dSYM       *)\n}\n```',
    'pre-build':     'Shell commands to run **before** compilation.\n\nEnvironment: `DOTORI_TARGET`, `DOTORI_CONFIG`, `DOTORI_PROJECT_DIR`, `DOTORI_OUTPUT_DIR`',
    'post-build':    'Shell commands to run **after** linking and artifact copy.\n\nEnvironment: `DOTORI_TARGET`, `DOTORI_CONFIG`, `DOTORI_PROJECT_DIR`, `DOTORI_OUTPUT_DIR`',

    // ── Dependency options ──────────────────────────────────────────────────
    'path':    'Local path dependency (included in the build DAG).',
    'git':     'Git repository URL for a dependency.',
    'tag':     'Git tag to check out for a git dependency.',
    'commit':  'Git commit hash for a git dependency.',
    'version': 'Version constraint for a dependency.',

    // ── Project types ───────────────────────────────────────────────────────
    'executable':      'Produces an executable binary.',
    'static-library':  'Produces a static library (`.a` / `.lib`).',
    'shared-library':  'Produces a shared library (`.so` / `.dll` / `.dylib`).',
    'header-only':     'Header-only library; no compilation step.',

    // ── Condition atoms ─────────────────────────────────────────────────────
    'windows':   'Condition: target is Windows (desktop).',
    'uwp':       'Condition: target is Universal Windows Platform.',
    'linux':     'Condition: target is Linux.',
    'android':   'Condition: target is Android (NDK).',
    'macos':     'Condition: target is macOS.',
    'ios':       'Condition: target is iOS.',
    'tvos':      'Condition: target is tvOS.',
    'watchos':   'Condition: target is watchOS.',
    'wasm':      'Condition: target is WebAssembly.',
    'debug':     'Condition: build configuration is Debug.',
    'release':   'Condition: build configuration is Release.',
    'msvc':      'Condition: compiler is MSVC.',
    'clang':     'Condition: compiler is Clang.',
    'static':    'Condition: runtime-link is static.',
    'dynamic':   'Condition: runtime-link is dynamic.',
    'glibc':     'Condition: libc is glibc.',
    'musl':      'Condition: libc is musl.',
    'libcxx':    'Condition: stdlib is libc++.',
    'libstdcxx': 'Condition: stdlib is libstdc++.',
    'emscripten':'Condition: WASM backend is Emscripten.',
    'bare':      'Condition: WASM backend is bare clang (wasm32-unknown-unknown).',

    // ── PCH fields ──────────────────────────────────────────────────────────
    'header':      'Precompiled header file path (`.h`).',
    'source':      'Precompiled header source file path (`.cpp`) — MSVC only.',

    // ── Unity-build fields ──────────────────────────────────────────────────
    'enabled':     'Enable unity build.\n\nValues: `true` | `false`',
    'batch-size':  'Number of source files per unity batch (default: `8`).',
    'exclude':     'Glob patterns for files to exclude from unity batching.',

    // ── Output fields ───────────────────────────────────────────────────────
    'binaries':   'Output directory for executables and shared libraries.',
    'libraries':  'Output directory for static libraries and Windows import libs.',
    'symbols':    'Output directory for debug symbols (`.pdb`, `.dSYM`).',

    // ── Package fields ──────────────────────────────────────────────────────
    'name':        'Package name.',
    'license':     'SPDX license identifier (e.g. `"MIT"`, `"Apache-2.0"`).',
    'homepage':    'Package homepage URL.',
    'authors':     'List of package authors.',
    'exports':     'Module or header path exports for consumers.',
    'export-map':  'Generate a module export map (`.json`) alongside BMI files.\n\nValues: `true` | `false` (default: `true`)',
};

export class DotoriHoverProvider implements vscode.HoverProvider {

    provideHover(
        document: vscode.TextDocument,
        position: vscode.Position,
    ): vscode.ProviderResult<vscode.Hover> {

        const wordRange = document.getWordRangeAtPosition(position, /[a-zA-Z][a-zA-Z0-9_\-]*/);
        if (!wordRange) { return undefined; }

        const word = document.getText(wordRange);
        const doc  = KEYWORD_DOCS[word];
        if (!doc) { return undefined; }

        const md = new vscode.MarkdownString();
        md.isTrusted = true;
        md.appendMarkdown(`**\`${word}\`** — ${doc}`);

        return new vscode.Hover(md, wordRange);
    }
}

File: src/Dotori.LanguageServer/Providers/HoverProvider.cs
using Dotori.LanguageServer.Protocol;

namespace Dotori.LanguageServer.Providers;

/// <summary>
/// Provides Markdown hover documentation for .dotori DSL keywords.
/// </summary>
public static class HoverProvider
{
    private static readonly Dictionary<string, string> Descriptions =
        new(StringComparer.OrdinalIgnoreCase)
        {
            ["type"]           = "프로젝트 빌드 타입 (executable, static-library, shared-library, header-only)",
            ["std"]            = "C++ 표준 버전 (c++17, c++20, c++23)",
            ["description"]    = "프로젝트 설명 문자열",
            ["runtime-link"]   = "런타임 링크 방식 (static: /MT, dynamic: /MD)",
            ["libc"]           = "C 런타임 라이브러리 (glibc, musl). Linux 타겟 전용",
            ["stdlib"]         = "C++ 표준 라이브러리 (libc++, libstdc++)",
            ["optimize"]       = "최적화 수준 (none: -O0, size: -Os, speed: -O2, full: -O3)",
            ["debug-info"]     = "디버그 정보 (none, minimal: -gline-tables-only, full: -g)",
            ["lto"]            = "링크 타임 최적화 (true/false). MSVC: /GL+/LTCG, Clang: -flto",
            ["warnings"]       = "경고 수준 (none, default, all, extra)",
            ["warnings-as-errors"] = "경고를 오류로 처리 (true/false). MSVC: /WX, Clang: -Werror",
            ["android-api-level"]  = "Android API 레벨 (예: 26). android-* 타겟 전용",
            ["macos-min"]      = "최소 macOS 버전 (예: \"12.0\")",
            ["ios-min"]        = "최소 iOS 버전 (예: \"15.0\")",
            ["tvos-min"]       = "최소 tvOS 버전",
            ["watchos-min"]    = "최소 watchOS 버전",
            ["pch"]            = "프리컴파일 헤더 설정 (header, source 경로 지정)",
            ["unity-build"]    = "Unity Build 설정 (여러 .cpp를 하나로 합쳐 컴파일 속도 향상)",
            ["dependencies"]   = "의존성 선언 (path 로컬 프로젝트, git/version 패키지)",
            ["sources"]        = "컴파일할 소스 파일 glob 패턴 (include/exclude)",
            ["modules"]        = "C++ Modules (.cppm/.ixx) 파일 glob 패턴",
            ["headers"]        = "헤더 경로 (public: 외부 노출, private: 내부 전용)",
            ["defines"]        = "전처리기 매크로 정의 (예: \"NDEBUG\" \"MY_DEFINE=1\")",
            ["links"]          = "링크할 라이브러리 이름 (예: \"pthread\" \"dl\")",
            ["frameworks"]     = "링크할 Apple 프레임워크 이름 (macOS/iOS 전용)",
            ["framework-paths"] = "Apple .framework/.xcframework 번들 경로",
            ["compile-flags"]  = "추가 컴파일러 플래그 (dotori 생성 플래그 뒤에 추가됨)",
            ["link-flags"]     = "추가 링커 플래그 (dotori 생성 플래그 뒤에 추가됨)",
            ["output"]         = "빌드 결과물 복사 경로 설정 (binaries, libraries, symbols)",
            ["pre-build"]      = "빌드 전 실행할 셸 명령어 목록",
            ["post-build"]     = "빌드 후 실행할 셸 명령어 목록",
            ["copy"]           = "파일/폴더를 지정 경로로 복사. `from \"glob\" to \"dest/\"` 형식. 변경된 파일만 복사 (증분). 조건 블록과 조합 가능",
            ["emscripten-flags"] = "Emscripten 전용 추가 플래그 (예: \"-sUSE_SDL=2\")",
            ["resources"]      = "Windows 리소스 파일 (.rc) 목록",
            ["manifest"]       = "Windows 앱 매니페스트 파일 (.manifest) 경로",
            ["option"]         = "선택적 빌드 옵션 선언. `default` (필수), `defines`, `dependencies` 지정 가능. `dotori build --옵션명`으로 활성화",
            // Package fields
            ["name"]           = "패키지 이름",
            ["version"]        = "패키지 버전 (예: \"1.0.0\")",
            ["license"]        = "라이선스 식별자 (예: \"MIT\", \"Apache-2.0\")",
            ["homepage"]       = "프로젝트 홈페이지 URL",
            ["authors"]        = "패키지 저자 목록",
            ["exports"]        = "패키지 내보내기 경로 맵",
            // Condition atoms
            ["windows"]    = "Windows 플랫폼 조건",
            ["linux"]      = "Linux 플랫폼 조건",
            ["macos"]      = "macOS 플랫폼 조건",
            ["ios"]        = "iOS 플랫폼 조건",
            ["tvos"]       = "tvOS 플랫폼 조건",
            ["watchos"]    = "watchOS 플랫폼 조건",
            ["android"]    = "Android 플랫폼 조건",
            ["uwp"]        = "UWP (Universal Windows Platform) 조건",
            ["wasm"]       = "WebAssembly 플랫폼 조건",
            ["debug"]      = "Debug 빌드 구성 조건",
            ["release"]    = "Release 빌드 구성 조건",
            ["msvc"]       = "MSVC (cl.exe / clang-cl) 컴파일러 조건",
            ["clang"]      = "Clang (clang++) 컴파일러 조건",
            ["static"]     = "Static 런타임 링크 조건",
            ["dynamic"]    = "Dynamic 런타임 링크 조건",
            ["glibc"]      = "glibc (GNU C Library) 런타임 조건",
            ["musl"]       = "musl libc 런타임 조건",
            ["libcxx"]     = "libc++ (LLVM C++ 표준 라이브러리) 조건",
            ["libstdcxx"]  = "libstdc++ (GCC C++ 표준 라이브러리) 조건",
            ["emscripten"] = "Emscripten WASM 백엔드 조건",
            ["bare"]       = "Bare (wasm32-unknown-unknown) WASM 백엔드 조건",
        };

    /// <summary>
    /// Get hover documentation for the word at the given position in the document.
    /// Returns null if no documentation is available.
    /// </summary>
    public static LspHover? GetHover(string text, int line, int character)
    {
        var lines = text.Split('\n');
        if (line >= lines.Length) return null;

        var currentLine = lines[line];
        var word = ExtractWordAt(currentLine, character);
        if (word is null) return null;

        if (!Descriptions.TryGetValue(word, out var description))
            return null;

        return new LspHover
        {
            Contents = new LspMarkupContent
            {
                Kind  = "markdown",
                Value = $"**{word}**\n\n{description}",
            },
            Range = GetWordRange(line, currentLine, character, word),
        };
    }

    private static string? ExtractWordAt(string line, int character)
    {
        if (character > line.Length) character = line.Length;

        // Find start of word (identifier chars: letters, digits, -, _)
        var start = character;
        while (start > 0 && IsWordChar(line[start - 1]))
            start--;

        // Find end of word
        var end = character;
        while (end < line.Length && IsWordChar(line[end]))
            end++;

        if (start >= end) return null;
        return line[start..end];
    }

    private static bool IsWordChar(char c) =>
        char.IsLetterOrDigit(c) || c == '-' || c == '_' || c == '+';

    private static LspRange GetWordRange(int line, string lineText, int character, string word)
    {
        var start = character;
        while (start > 0 && IsWordChar(lineText[start - 1]))
            start--;

        return new LspRange
        {
            Start = new LspPosition { Line = line, Character = start },
            End   = new LspPosition { Line = line, Character = start + word.Length },
        };
    }
}
